imports/wasi_snapshot_preview1/wasi_stdlib_test.go

```diff
@@ -396,3 +402,67 @@ func testSock(t *testing.T, bin []byte) {
 	require.NoError(t, err)
 	require.Equal(t, "wazero\n", console)
 }
+
+func Test_HTTP(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("syscall.Nonblocking() is not supported on wasip1+windows.")
+	}
+	toolchains := map[string][]byte{}
+	if wasmGotip != nil {
+		toolchains["gotip"] = wasmGotip
+	}
+
+	for toolchain, bin := range toolchains {
+		toolchain := toolchain
+		bin := bin
+		t.Run(toolchain, func(t *testing.T) {
+			testHTTP(t, bin)
+		})
+	}
+}
+
+func testHTTP(t *testing.T, bin []byte) {
+	sockCfg := experimentalsock.NewConfig().WithTCPListener("127.0.0.1", 0)
+	ctx := experimentalsock.WithConfig(testCtx, sockCfg)
+	ctx, cancelFunc := context.WithCancel(ctx)
+	// Set context to one that has an experimental listener that logs all host functions.
+	// ctx = context.WithValue(ctx, experimental.FunctionListenerFactoryKey{},
+	//	logging.NewHostLoggingListenerFactory(os.Stdout, logging.LogScopeAll))
+
+	moduleConfig := wazero.NewModuleConfig().
+		WithSysWalltime().WithSysNanotime(). // HTTP middleware uses both clocks
+		WithArgs("wasi", "http")
+	tcpAddrCh := make(chan *net.TCPAddr, 1)
+	ch := make(chan string, 1)
+	go func() {
+		ch <- compileAndRunWithPreStart(t, ctx, moduleConfig, bin, func(t *testing.T, mod api.Module) {
+			tcpAddrCh <- requireTCPListenerAddr(t, mod)
+		})
+	}()
+	tcpAddr := <-tcpAddrCh
+
+	// Give a little time for _start to complete
+	sleepALittle()
+
+	// Now, send a POST to the address which we had pre-opened.
+	body := bytes.NewReader([]byte("wazero"))
+	req, err := http.NewRequest(http.MethodPost, "http://"+tcpAddr.String(), body)
+	require.NoError(t, err)
+
+	// TODO: test hangs here
+	resp, err := http.DefaultClient.Do(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+
+	require.Equal(t, 200, resp.StatusCode)
+	b, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	require.Equal(t, "wazero\n", string(b))
+
+	cancelFunc()
+
+	// FIXME: this hangs because the example listens forever;
+	// 	      however mod.Close() does not result in a clean shudown.
+	// console := <-ch
+	// require.Equal(t, "", console)
```

Comment:
solving by serving only 1 request and then exiting the Wasm routine

Reply:
hah I was thinking about this route also.